frame/alliance/src/lib.rs

```diff
@@ -0,0 +1,1049 @@
+			T::AnnouncementOrigin::ensure_origin(origin)?;
+
+			let mut announcements = <Announcements<T, I>>::get();
+			announcements
+				.try_push(announcement.clone())
+				.map_err(|_| Error::<T, I>::TooManyAnnouncements)?;
+			<Announcements<T, I>>::put(announcements);
+
+			Self::deposit_event(Event::Announced { announcement });
+			Ok(())
+		}
+
+		/// Remove an announcement.
+		#[pallet::weight(T::WeightInfo::remove_announcement())]
+		pub fn remove_announcement(origin: OriginFor<T>, announcement: Cid) -> DispatchResult {
+			T::AnnouncementOrigin::ensure_origin(origin)?;
+
+			let mut announcements = <Announcements<T, I>>::get();
+			let pos = announcements
+				.binary_search(&announcement)
+				.ok()
+				.ok_or(Error::<T, I>::MissingAnnouncement)?;
+			announcements.remove(pos);
+			<Announcements<T, I>>::put(announcements);
+
+			Self::deposit_event(Event::AnnouncementRemoved { announcement });
+			Ok(())
+		}
+
+		/// Submit oneself for candidacy. A fixed deposit is reserved.
+		#[pallet::weight(T::WeightInfo::join_alliance())]
+		pub fn join_alliance(origin: OriginFor<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// Blacklisted accounts cannot join.
+			ensure!(!Self::account_blacklisted(&who), Error::<T, I>::AlreadyInBlacklist);
+			ensure!(!Self::is_member(&who), Error::<T, I>::AlreadyMember);
+			// check user self or parent should has verified identity to reuse display name and
+			// website.
+			Self::has_identity(&who)?;
+
+			let deposit = T::AllyDeposit::get();
+			T::Currency::reserve(&who, deposit).map_err(|_| Error::<T, I>::InsufficientFunds)?;
+			<DepositOf<T, I>>::insert(&who, deposit);
+
+			Self::add_member(&who, MemberRole::Ally)?;
+
+			Self::deposit_event(Event::NewAllyJoined {
+				ally: who,
+				nominator: None,
+				reserved: Some(deposit),
+			});
+			Ok(())
+		}
+
+		/// A founder or fellow can nominate someone to join the alliance as an Ally.
+		/// There is no deposit required to the nominator or nominee.
+		#[pallet::weight(T::WeightInfo::nominate_ally())]
+		pub fn nominate_ally(
+			origin: OriginFor<T>,
+			who: <T::Lookup as StaticLookup>::Source,
+		) -> DispatchResult {
+			let nominator = ensure_signed(origin)?;
+			ensure!(Self::has_voting_rights(&nominator), Error::<T, I>::NoVotingRights);
+			let who = T::Lookup::lookup(who)?;
+
+			// Individual voting members cannot nominate blacklisted accounts.
+			ensure!(!Self::account_blacklisted(&who), Error::<T, I>::AlreadyInBlacklist);
+			ensure!(!Self::is_member(&who), Error::<T, I>::AlreadyMember);
+			// check user self or parent should has verified identity to reuse display name and
+			// website.
+			Self::has_identity(&who)?;
+
+			Self::add_member(&who, MemberRole::Ally)?;
+
+			Self::deposit_event(Event::NewAllyJoined {
+				ally: who,
+				nominator: Some(nominator),
+				reserved: None,
+			});
+			Ok(())
+		}
+
+		/// Elevate an ally to fellow.
+		#[pallet::weight(T::WeightInfo::elevate_ally())]
```

Comment:
Just wondering if this might be a good place for `#[transactional]` ?

Reply:
Follow up as part of the effort to make pallets transactional?

Reply:
Pallets are now transactional by default.